### PR TITLE
Add policy for '-' and '_' to scripts/check_ifndefs.py

### PR DESCRIPTION
**Pull request by scober**
I noticed that scripts/check_ifndefs.py, when constructing a header guard, does not handle some characters that may be valid in the path name but invalid (`-`) or misleading (`_`) in the include guard. '_' and '-' seemed like the most common characters to address, but there are other characters that would be valid in a path name but not in an in `#ifdef`.

### CODE REVIEW

**Comment by scober** [comment]
It turns out '_' was ignored on purpose.

**Comment by ddn0** [comment]
FWIW, `_` => "" would be fine too, you just need to run `scripts/check_ifndef.py -fix lib*` as well.

**Comment by scober** [comment]
> FWIW, `_` => "" would be fine too, you just need to run `scripts/check_ifndef.py -fix lib*` as well.

As the original author, do you have a preference?

**Comment by tylershunt** [comment]
If we were to handle `"_"` in a special way I would rather make it harder to shadow files, for instance turning it into `"__"` (double underscore) rather than striping away the difference between camel case and snake case.

**Comment by scober** [comment]
I like that.

**Comment by ddn0** [comment]
The original reason I didn't handle `_` (specially) was because I didn't want collisions between guards of different files. Keeping the mapping one-to-one guarantees that.

Escaping is the correct answer, though it looks kind of ugly in practice. (At some point, baseXY encoding the path becomes a viable answer too.)

- `_` and `-` => "" is reasonable because it is simple to explain and remember
- `-` => "", `_` => `_`  also seems simple enough
- `_` => `__` seems more borderline to me. At this point, I begin to think `#pragma once` is also okay. Since this is automated, the pain of changing policies is low, so I don't have a strong opposition to this choice either.

**Comment by ddn0** [comment]
Also, technically
- the classes `LC_CSR_XYZ` are going to be refactored with new property graph support anyway, and
- `Executor_ForEach.h` and co could equally be called `ExecutorForEach.h`
So `_` can be less common in names.

File: scripts/check_ifndefs.py [hunk 1]
```diff
@@ -49,6 +49,10 @@ def make_guard(root, filename):
     p = p.replace("/SRC/", "/", 1)
     p = p.replace("/TESTS/", "/", 1)
     p = p.replace("/TEST/", "/", 1)
+    # Just in case, remove characters that can't be part of macros
+    p = re.sub("[+\-*%=<>?~&\^|#:;{}.[\]]","", p)
+    # Differentiate between snake_case file names and directories
+    p = p.replace("_", "__", 1)
```
Comment (ddn0) [inline scripts/check_ifndefs.py:55]: `p.replace(..., -1)` to replace all occurrences?

**Comment by tylershunt** [comment]
> The original reason I didn't handle `_` (specially) was because I didn't want collisions between guards of different files. Keeping the mapping one-to-one guarantees that.

That makes sense as a goal but I would stop short of saying that the identity mapping prevents collisions. For instance the current file `libtsuba/include/tsuba/s3_internal.h` would collide with `libtsuba/include/tsuba/s3/internal.h`. That seems ok since we don't create deep nesting in the headers. What I was more concerned about were cases like `libtsuba/src/tsuba_internal.h` which (I think) collides with `libtsuba/include/tsuba/internal.h`

Yeah I take your point about `pragma once`

**Comment by scober** [comment]
Is the concern about `#pragma once` that we want to support building in environments that don't offer it?

**Comment by scober** [comment]
I think I like `_` => "" because "don't have two files in the same directory with the same name with one in camel case and one in snake case" seems like a reasonable policy and would eliminate the only(?) collision.

**Comment by ddn0** [comment]
I think `#pragma once` is just us old grognards wondering why we need all these new fangled things when back in our day we used m4 and liked it. And technically it is not standardized and future C++ module support will probably supersede it, so inertia might be beneficial here.

I don't think we have a concrete platforms that we intend to support that don't also support `#pragma once` too.

**Pull request merged**